src/main/host/syscall/type_formatting.rs

```diff
@@ -6,6 +6,30 @@ use crate::host::memory_manager::MemoryManager;
 use crate::host::syscall::io::read_sockaddr;
 use crate::host::syscall_types::ForeignArrayPtr;
 
+fn fmt_int_with_suffix(
+    f: &mut std::fmt::Formatter<'_>,
+    ptr: u64,
+    suffix: &str,
+) -> std::fmt::Result {
+    if ptr == 0 {
+        write!(f, "{ptr:#x} <null>")
+    } else {
+        write!(f, "{ptr:#x} {}", suffix)
+    }
+}
+
+fn fmt_ptr_with_suffix<T>(
+    f: &mut std::fmt::Formatter<'_>,
+    ptr: ForeignPtr<T>,
+    suffix: &str,
+) -> std::fmt::Result {
+    if ptr.is_null() {
+        write!(f, "{ptr:p} <null>")
+    } else {
+        write!(f, "{ptr:p} {}", suffix)
+    }
+}
+
```

Comment:
Should these be macros since they are called from macros?

Reply:
A function is better here I think.